### PR TITLE
fix: don't count initial message in forum threads towards `message_count`

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

Followup to #730.
It appears that the message type of the initial forum thread message isn't `thread_starter_message (21)`, but just `default (0)`.
We relied on the message type to accurately update the message count fields, which means it's now off by one for newly created forum threads.

This PR instead mirrors the internal client behavior by also checking the message ID, see the comments in the code.

## Checklist

- [x] If code changes were made, then they have been tested
    - [ ] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [x] This PR fixes an issue
- [ ] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: disnake/state.py [hunk 1]
```diff
@@ -784,24 +784,37 @@ def parse_message_create(self, data: gateway.MessageCreateEvent) -> None:
             self._messages.append(message)
         # we ensure that the channel is a type that implements last_message_id
         if channel:
-            if channel.__class__ in (TextChannel, Thread, VoiceChannel):
+            if type(channel) in (TextChannel, Thread, VoiceChannel):
                 channel.last_message_id = message.id  # type: ignore
-            if (
-                channel.__class__ is Thread
-                and message.type is not MessageType.thread_starter_message
+            # woo fun logic
```
Comment (onerandomusername) [inline disnake/state.py:789]: unecessary comment

File: disnake/state.py [hunk 2]
```diff
@@ -784,24 +784,37 @@ def parse_message_create(self, data: gateway.MessageCreateEvent) -> None:
             self._messages.append(message)
         # we ensure that the channel is a type that implements last_message_id
         if channel:
-            if channel.__class__ in (TextChannel, Thread, VoiceChannel):
+            if type(channel) in (TextChannel, Thread, VoiceChannel):
```
Comment (onerandomusername) [inline disnake/state.py:787]: keep this as `__class__`

File: disnake/state.py [hunk 3]
```diff
@@ -784,24 +784,37 @@ def parse_message_create(self, data: gateway.MessageCreateEvent) -> None:
...
                 channel.last_message_id = message.id  # type: ignore
-            if (
-                channel.__class__ is Thread
-                and message.type is not MessageType.thread_starter_message
+            # woo fun logic
+            # Essentially, messages *don't* count towards message_count, if:
+            # - they're the thread starter message
+            # - or, they're the initial message of a forum channel thread (which uses MessageType.default)
+            # This mirrors the current client and API behavior.
+            if type(channel) is Thread and not (
```
Comment (onerandomusername) [inline disnake/state.py:794]:
```suggestion
            if channel.__class__ is Thread and not (
```

File: disnake/state.py [hunk 4]
```diff
@@ -819,7 +832,11 @@ def parse_message_delete_bulk(self, data: gateway.MessageDeleteBulkEvent) -> Non
         guild = self._get_guild(raw.guild_id)
         thread = guild and guild.get_thread(raw.channel_id)
         if thread:
-            thread.message_count = max(0, thread.message_count - len(raw.message_ids))
+            to_subtract = len(raw.message_ids)
+            # the initial message isn't counted, and hence shouldn't be subtracted from the count either
```
Comment (onerandomusername) [inline disnake/state.py:836]: because we don't check the message type here is thread_starter_message, I wonder if we need to check for thread_starter_message above too.

**Review by onerandomusername — CHANGES_REQUESTED** [review]

File: disnake/state.py (hunk 2 above)
Comment (shiftinv) [reply disnake/state.py:787]: This change was to avoid two additional `# type: ignore`s, since pyright doesn't support narrowing on `.__class__` :/ (see https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md#type-guards). The difference in terms of performance is negligible (a few nanoseconds), so imo this is worth it, but I'll open a pyright issue in hopes of getting this changed in the future.

File: disnake/state.py (hunk 1 above)
Comment (shiftinv) [reply disnake/state.py]: https://github.com/DisnakeDev/disnake/pull/747/commits/f909613e726ca56859c18267390660a1c4806bc7

File: disnake/state.py (hunk 4 above)
Comment (shiftinv) [reply disnake/state.py:831]: It should be safe to assume that if `msg.type == thread_starter_message`, then `msg.id == ch.id`, so the check in `parse_message_create` could possibly be simplified to just `msg.id == ch.id`.
The current implementation in all three event handlers mirrors the client, which checks type/parent/id in `message_create`, and just the id in `message_delete` and `message_delete_bulk`, so I'm not sure - either option should be fine.

File: disnake/state.py (hunk 2 above)
Comment (onerandomusername) [reply disnake/state.py:787]: image

File: disnake/state.py (hunk 4 above)
Comment (onerandomusername) [reply disnake/state.py:831]: That's fair. We probably don't need to check thread_starter_message in the message create but I don't know.

Comment (shiftinv) [reply disnake/state.py:831]: Nevermind, I mixed up the IDs, this is wrong:

> It should be safe to assume that if `msg.type == thread_starter_message`, then `msg.id == ch.id`

Meaning the message type check is still required for non-forum threads.

Comment (onerandomusername) [reply disnake/state.py:831]: Ah.

File: disnake/state.py [hunk 5]
```diff
@@ -784,24 +784,32 @@ def parse_message_create(self, data: gateway.MessageCreateEvent) -> None:
...
+            if type(channel) in (TextChannel, Thread, VoiceChannel):
                 channel.last_message_id = message.id  # type: ignore
-            if (
-                channel.__class__ is Thread
-                and message.type is not MessageType.thread_starter_message
+            # Essentially, messages *don't* count towards message_count, if:
+            # - they're the thread starter message
+            # - or, they're the initial message of a forum channel thread (which uses MessageType.default)
+            # This mirrors the current client and API behavior.
+            if type(channel) is Thread and not (
```
Comment (EQUENOS) [inline disnake/state.py:793]: Why not `isinstance(channel, Thread)`?

Comment (Snipy7374) [reply disnake/state.py:793]: FYI we had a discussion about this before
https://discord.com/channels/808030843078836254/913779868985090089/1017172795039432786

Comment (EQUENOS) [reply disnake/state.py:793]: Why do we want this type check to be strict? Do we have any `Thread` subclasses to worry about this?
Thing is, `isinstance(...)` seems to be almost 1.5 times faster than `type(...) is ...` so that's why I'm asking.

Comment (shiftinv) [reply disnake/state.py:793]: > Thing is, `isinstance(...)` seems to be almost 1.5 times faster than `type(...) is ...` so that's why I'm asking.

I can't seem to replicate that:
```py
In [1]: from disnake import TextChannel, Thread

In [2]: textchannel = TextChannel.__new__(TextChannel)

In [3]: thread = Thread.__new__(Thread)

In [4]: %timeit isinstance(thread, Thread)
56.3 ns ± 2.05 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [5]: %timeit type(thread) is Thread
56.4 ns ± 1.12 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [6]: %timeit isinstance(textchannel, Thread)
75.1 ns ± 2.15 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [7]: %timeit type(textchannel) is Thread
56.8 ns ± 1.5 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

Comment (Snipy7374) [reply disnake/state.py:793]: I was already aware that my measurements were not reliable, most likely the speed depends on the cases, have you also tried to measure the execution time with a real Thread? (I searched on the internet in numerous forums and generally, through some measures with the timeit module, the fastest solution is `isinstance`)

Comment (onerandomusername) [reply disnake/state.py:793]: @shiftinv what times do you get for comparing `__class__` as well?

Comment (shiftinv) [reply disnake/state.py:793]: Let's do it properly then, if we're counting nanoseconds :^)
~~`pyperf --duplicate` my beloved~~


```py
$ python test.py -q
__class__ (no match): Mean +- std dev: 27.1 ns +- 2.3 ns
__class__ (match): Mean +- std dev: 27.6 ns +- 4.6 ns
type (no match): Mean +- std dev: 37.0 ns +- 4.0 ns
type (match): Mean +- std dev: 36.7 ns +- 3.4 ns
isinstance (no match): Mean +- std dev: 56.2 ns +- 4.5 ns
isinstance (match): Mean +- std dev: 35.1 ns +- 3.0 ns
```
stddev is pretty high at around 10% of the mean, but comparing these should still be relatively accurate

```py
import pyperf

runner = pyperf.Runner()

setup = """
from disnake import TextChannel, Thread
c = TextChannel.__new__(TextChannel)
t = Thread.__new__(Thread)
"""

kwargs = {"setup": setup, "duplicate": 512}

runner.timeit("__class__ (no match)", "c.__class__ is Thread", **kwargs)
runner.timeit("__class__ (match)", "t.__class__ is Thread", **kwargs)
runner.timeit("type (no match)", "type(c) is Thread", **kwargs)
runner.timeit("type (match)", "type(t) is Thread", **kwargs)
runner.timeit("isinstance (no match)", "isinstance(c, Thread)", **kwargs)
runner.timeit("isinstance (match)", "isinstance(t, Thread)", **kwargs)
```

tldr `isinstance` if the type doesn't match is the slowest, everything else doesn't matter in practice

**Review by onerandomusername — APPROVED** [review]
